pkg/server/push/cert_apns.go

```diff
@@ -0,0 +1,213 @@
+		x509Cert, err := x509.ParseCertificate(cert)
+		if err != nil {
+			return err
+		}
+		certificate.Leaf = x509Cert
+		return nil
+	}
+
+	return errors.New("push/apns: provided APNS certificate does not contain leaf")
+}
+
+// findDefaultAPNSTopic returns the APNS topic in the TLS certificate.
+//
+// The Subject of leaf certificate should contains the UID, which we can
+// use as the topic for APNS. The topic is usually the same as the
+// application bundle // identifier.
+//
+// Returns the topic name, and an error if an error occurring finding the topic
+// name.
+func findDefaultAPNSTopic(certificate tls.Certificate) (string, error) {
+	if certificate.Leaf == nil {
+		err := parseCertificateLeaf(&certificate)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Loop over the subject names array to look for UID
+	uidObjectIdentifier := asn1.ObjectIdentifier([]int{0, 9, 2342, 19200300, 100, 1, 1})
+	for _, attr := range certificate.Leaf.Subject.Names {
+		if uidObjectIdentifier.Equal(attr.Type) {
+			switch value := attr.Value.(type) {
+			case string:
+				return value, nil
+			}
+			break
+		}
+	}
+
+	return "", errors.New("push/apns: cannot find UID in APNS certificate subject name")
+}
+
+// NewCertBasedAPNSPusher returns a new APNSPusher from content of certificate
+// and private key as string
+func NewCertBasedAPNSPusher(
+	connOpener func() (skydb.Conn, error),
+	gatewayType GatewayType,
+	cert string,
+	key string,
+) (APNSPusher, error) {
+	certificate, err := tls.X509KeyPair([]byte(cert), []byte(key))
+	if err != nil {
+		return nil, err
+	}
+
+	topic, err := findDefaultAPNSTopic(certificate)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := push.NewClient(certificate)
+	if err != nil {
+		return nil, err
+	}
+
+	service, err := newPushService(client, gatewayType)
+	if err != nil {
+		return nil, err
+	}
+
+	return &certBasedAPNSPusher{
+		connOpener: connOpener,
+		service:    service,
+		topic:      topic,
+	}, nil
+}
+
+// Start setups the pusher and starts it
+func (pusher *certBasedAPNSPusher) Start() {
+	conn, err := pusher.connOpener()
+	if err != nil {
+		log.Errorf("push/apns: failed to open skydb.Conn, abort feedback retrival: %v\n", err)
+		return
+	}
+
+	pusher.conn = conn
+	pusher.failed = make(chan failedNotification)
+
+	go func() {
+		checkFailedNotifications(pusher)
+	}()
+}
+
+// Stop stops and cleans up the pusher
+func (pusher *certBasedAPNSPusher) Stop() {
+	close(pusher.failed)
+	pusher.failed = nil
+}
+
+// Send sends a notification to the device identified by the
+// specified device
+func (pusher *certBasedAPNSPusher) Send(m Mapper, device skydb.Device) error {
+	logger := log.WithFields(logrus.Fields{
+		"deviceToken": device.Token,
+		"deviceID":    device.ID,
+		"deviceTopic": device.Topic,
+	})
+
+	if m == nil {
+		logger.Warn("Cannot send push notification with nil data.")
+		return errors.New("push/apns: push notification has no data")
+	}
+
+	if device.Topic == "" {
```

Comment:
Since cert-based pusher does not require topic (the topic is in the certificate), what will happen if the device has null topic here?